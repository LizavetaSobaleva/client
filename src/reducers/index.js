import { applyMiddleware, combineReducers } from "redux";
import { configureStore} from "@reduxjs/toolkit"
import { thunk } from "redux-thunk";
import userReducer from "./userReducer";
import fileReducer from "./fileReducer";
import uploadReducer from "./uploadReducer";
import appReducer from "./appReducer";
import { composeWithDevTools } from '@redux-devtools/extension';


const rootReducer = combineReducers({
    user: userReducer,
    files: fileReducer,
    upload: uploadReducer,
    app: appReducer
})

export const store = configureStore({reducer: rootReducer, devTools: composeWithDevTools(applyMiddleware(thunk))})