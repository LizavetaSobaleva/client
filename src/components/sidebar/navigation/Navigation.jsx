import React, { useState } from 'react'
import './navigation.less'
import DirItem from './DirItem'
import { useDispatch, useSelector } from 'react-redux'
import { clearStack, setCurrentDir } from '../../../reducers/fileReducer'
import { NavLink } from 'react-router-dom'
import Dir from '../../../assets/img/dir.svg'
import Star from '../../../assets/img/star.svg'
import Settings from '../../../assets/img/settings.svg'
import DropdownButton from '../../UI/dropdownButton/DropdownButton'


const Navigation = () => {
    const dispatch = useDispatch()
    const [isOpen, setIsOpen] = useState(false)
    const dirStructure = useSelector((state) => state.files.dirStructure)
    const hasChildren = dirStructure && dirStructure.length > 0


    function openHandler() {
        dispatch(clearStack())
        dispatch(setCurrentDir(null))
    }

    const toggleDropdown = () => {
      setIsOpen(!isOpen)
    }
    

  return (
    <div className='navigation'>
      
      <div className="navigation__item">
      <NavLink to="/"  onClick={() => openHandler()}>
        <div className="navigation__title">
          <img src={Dir} alt="" className="navigation__icon"/>
          <div className="navigation__name">My Cloud</div>

          {hasChildren && (
            <div className="navigation__dropdown" onClick={toggleDropdown}>
              <DropdownButton />
            </div>
          )}
        </div>
        </NavLink>
        <div className={`navigation__content ${isOpen ? 'visible' : ''}`} >
          {hasChildren && (
            <div className="navigation__element">
              {dirStructure.map((dir) => 
                <DirItem file={dir} title={dir.name} key={dir.path}/>
              )}
            </div>
          )}
        </div>
      </div>
      

      <div className="navigation__item">
      <div className="navigation__title">
          <img src={Star} alt="" className="navigation__icon"/>
          <div className="navigation__name">Menu Item 2</div>
        </div>
      </div>

      <div className="navigation__item">
      <div className="navigation__title">
          <img src={Settings} alt="" className="navigation__icon"/>
          <div className="navigation__name">Menu Item 3</div>
        </div>
      </div>
        
        
        
    </div>
  )
}

export default Navigation