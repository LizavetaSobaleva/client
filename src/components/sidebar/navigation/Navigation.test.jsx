import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import Navigation from './Navigation';
import { clearStack, setCurrentDir } from '../../../reducers/fileReducer';

const mockStore = configureStore([]);
let store;

const renderComponent = (store) =>
  render(
    <Provider store={store}>
      <MemoryRouter>
        <Navigation />
      </MemoryRouter>
    </Provider>
  );

describe('Navigation Component', () => {
  beforeEach(() => {
    store = mockStore({
      files: { dirStructure: [{ _id: '1', name: 'Dir 1', path: 'dir1', idPath: ['1'], childs: [] }] }
    });
    store.dispatch = jest.fn();
  });

  test('renders Navigation without crashing', () => {
    const { asFragment } = renderComponent(store);
    expect(screen.getByTestId('navigation')).toBeInTheDocument();
    expect(screen.getByTestId('navigationItem')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders navigation title and icon', () => {
    const { asFragment } = renderComponent(store);
    expect(screen.getByTestId('navigationTitle')).toHaveTextContent('My Cloud');
    expect(screen.getByTestId('navigationIcon')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('opens and closes dropdown correctly', () => {
    renderComponent(store);
    const dropdown = screen.getByTestId('navigationDropdown');
    const content = screen.getByTestId('navigationContent');
    expect(content).toHaveClass('visible');
    fireEvent.click(dropdown);
    expect(content).not.toHaveClass('visible');
    fireEvent.click(dropdown);
    expect(content).toHaveClass('visible');
  });

  test('clicking My Cloud link triggers correct actions', () => {
    renderComponent(store);
    const navLink = screen.getByTestId('navLink');
    fireEvent.click(navLink);
    expect(store.dispatch).toHaveBeenCalledWith(clearStack());
    expect(store.dispatch).toHaveBeenCalledWith(setCurrentDir(null));
  });

  test('renders directory structure correctly', () => {
    const { asFragment } = renderComponent(store);
    expect(screen.getByTestId('navigationElement')).toBeInTheDocument();
    expect(screen.getByText('Dir 1')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
