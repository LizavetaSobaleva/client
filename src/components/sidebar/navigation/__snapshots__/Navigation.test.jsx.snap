// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation Component renders Navigation without crashing 1`] = `
<DocumentFragment>
  <div
    class="navigation"
    data-testid="navigation"
  >
    <div
      class="navigation__item"
      data-testid="navigationItem"
    >
      <a
        aria-current="page"
        class="active"
        data-testid="navLink"
        href="/"
      >
        <div
          class="navigation__title"
          data-testid="navigationTitle"
        >
          <img
            alt=""
            class="navigation__icon"
            data-testid="navigationIcon"
            src="test-file-stub"
          />
          <div
            class="navigation__name"
          >
            My Cloud
          </div>
          <div
            class="navigation__dropdown"
            data-testid="navigationDropdown"
          >
            <div
              class="dropdown"
              data-testid="dropdownButton"
            >
              <img
                alt=""
                class="dropdown__icon "
                data-testid="dropdownIcon"
                src="test-file-stub"
              />
            </div>
          </div>
        </div>
      </a>
      <div
        class="navigation__content visible"
        data-testid="navigationContent"
      >
        <div
          class="navigation__element"
          data-testid="navigationElement"
        >
          <div
            class="diritem"
            data-testid="dirItem-1"
          >
            <div
              class="diritem__content"
            >
              <div
                class="diritem__title"
                data-testid="dirItemTitle-1"
              >
                Dir 1
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Navigation Component renders directory structure correctly 1`] = `
<DocumentFragment>
  <div
    class="navigation"
    data-testid="navigation"
  >
    <div
      class="navigation__item"
      data-testid="navigationItem"
    >
      <a
        aria-current="page"
        class="active"
        data-testid="navLink"
        href="/"
      >
        <div
          class="navigation__title"
          data-testid="navigationTitle"
        >
          <img
            alt=""
            class="navigation__icon"
            data-testid="navigationIcon"
            src="test-file-stub"
          />
          <div
            class="navigation__name"
          >
            My Cloud
          </div>
          <div
            class="navigation__dropdown"
            data-testid="navigationDropdown"
          >
            <div
              class="dropdown"
              data-testid="dropdownButton"
            >
              <img
                alt=""
                class="dropdown__icon "
                data-testid="dropdownIcon"
                src="test-file-stub"
              />
            </div>
          </div>
        </div>
      </a>
      <div
        class="navigation__content visible"
        data-testid="navigationContent"
      >
        <div
          class="navigation__element"
          data-testid="navigationElement"
        >
          <div
            class="diritem"
            data-testid="dirItem-1"
          >
            <div
              class="diritem__content"
            >
              <div
                class="diritem__title"
                data-testid="dirItemTitle-1"
              >
                Dir 1
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Navigation Component renders navigation title and icon 1`] = `
<DocumentFragment>
  <div
    class="navigation"
    data-testid="navigation"
  >
    <div
      class="navigation__item"
      data-testid="navigationItem"
    >
      <a
        aria-current="page"
        class="active"
        data-testid="navLink"
        href="/"
      >
        <div
          class="navigation__title"
          data-testid="navigationTitle"
        >
          <img
            alt=""
            class="navigation__icon"
            data-testid="navigationIcon"
            src="test-file-stub"
          />
          <div
            class="navigation__name"
          >
            My Cloud
          </div>
          <div
            class="navigation__dropdown"
            data-testid="navigationDropdown"
          >
            <div
              class="dropdown"
              data-testid="dropdownButton"
            >
              <img
                alt=""
                class="dropdown__icon "
                data-testid="dropdownIcon"
                src="test-file-stub"
              />
            </div>
          </div>
        </div>
      </a>
      <div
        class="navigation__content visible"
        data-testid="navigationContent"
      >
        <div
          class="navigation__element"
          data-testid="navigationElement"
        >
          <div
            class="diritem"
            data-testid="dirItem-1"
          >
            <div
              class="diritem__content"
            >
              <div
                class="diritem__title"
                data-testid="dirItemTitle-1"
              >
                Dir 1
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
